### ISSUE TITLE
Improved types for Functions

### ISSUE THREAD

**Issue by knightedcodemonkey** [oxc-project/oxc#10988]
Currently the `Function` interface has a `type` defined as

```ts
export type FunctionType =
  | 'FunctionDeclaration'
  | 'FunctionExpression'
  | 'TSDeclareFunction'
  | 'TSEmptyBodyFunctionExpression';
```

which requires `Function.body` to be of type `FunctionBody | null` since a `TSDeclareFunction` can have a `null` body, for example when defining overload signatures

```ts
function foo(): void
function foo(arg0?: string): void { }
```

or a `TSEmptyBodyFunctionExpression`

```ts
const f = function ()
```

Grouping all the different function types together under `Function` leads to situations where narrowing is required to satisfy the TS compiler to ensure the existence of a `body` even when a body is always required by the spec, for example in a `FunctionExpression`

```ts
switch (func.type) {
  case 'FunctionExpression':
    func.body.start // <--- 'func.body' is possibly 'null' TS error
```

Ideally the types are defined so that after narrowing the node to a `FunctionExpression` or `FunctionDeclaration` the compiler doesn't warn about a possible `null` body. @camc314 suggested a possible approach for this on #9128: https://github.com/oxc-project/oxc/pull/9128#issuecomment-287022046

**Comment by overlookmotel** [comment]
My feeling is that ideal solution in long run is to have separate Rust types for `FunctionExpression`, `FunctionDeclaration`, `TSDeclareFunction` etc. We have the same kind of problem as described here on Rust side too with `Function` - we're constantly having to do `func.body.unwrap()` where we know the function has a body because it's an actual function not a TS declaration, which is error-prone and unperformant. `FunctionExpression` and `FunctionDeclaration` also differ from each other in where their `id` symbol is bound. Again we have "workaround" code to handle this.

But... that'd be a major change, and isn't going to happen any time soon, so in meantime we need to treat the symptom rather than the cause and amend the TS types.

The problem is that I'm absolutely rubbish at TypeScript, so I'm not sure how to tackle that.

@camc314 can I ask for your help with this? If you could describe what we ideally would like the TS type defs to look like, I can try to figure out how to make the codegen produce that.

Is the ideal from consumer's perspective to have separate `FunctionExpression`, `FunctionDeclaration`, and `TSDeclareFunction` TS types, and remove the `Function` type entirely? Or are type assertions (or something) preferable?

**Comment by camc314** [comment]
> My feeling is that ideal solution in long run is to have separate Rust types for FunctionExpression, FunctionDeclaration, TSDeclareFunction etc. We have the same kind of problem as described here on Rust side too with Function - we're constantly having to do func.body.unwrap() where we know the function has a body because it's an actual function not a TS declaration, which is error-prone and unperformant. FunctionExpression and FunctionDeclaration also differ from each other in where their id symbol is bound. Again we have "workaround" code to handle this.

In my opinion, the rust types are good enough, adding in a 3rd type, `TSFunction` or similar (imo) just adds a big more complexity (maybe not, its's actually all trade offs i guess)

https://github.com/oxc-project/oxc/blob/1bc8d292090d31c6537fe105652ee4d8813d3308/crates/oxc_ast/src/ast/js.rs#L1693-L1694

https://github.com/oxc-project/oxc/blob/1bc8d292090d31c6537fe105652ee4d8813d3308/crates/oxc_ast/src/ast/js.rs#L1880-L1899

But... for the TS type, `Function` should at the minium be split into two
 
<collapse>

In the collapse is the diff of `npm/oxc-types/types.d.ts` of what exactly i would change the types to. which should provide good typesafety to whoever is using this
<details>

```diff
diff --git a/npm/oxc-types/types.d.ts b/npm/oxc-types/types.d.ts
index 1e52c1936..9e6d8bdd4 100644
--- a/npm/oxc-types/types.d.ts
+++ b/npm/oxc-types/types.d.ts
@@ -28,7 +28,7 @@ export type Expression =
   | ChainExpression
   | Class
   | ConditionalExpression
-  | Function
+  | FunctionExpression
   | ImportExpression
   | LogicalExpression
   | NewExpression
@@ -576,26 +576,41 @@ export interface BindingRestElement extends Span {
   value?: null;
 }
 
-export interface Function extends Span {
-  type: FunctionType;
+export type Function =
+  & (Span & {
+    id: BindingIdentifier; // must be defined for all non-expression functions
+
+    generator: boolean;
+    async: boolean;
+    params: ParamPattern[];
+    typeParameters?: TSTypeParameterDeclaration | null;
+    returnType?: TSTypeAnnotation | null;
+  })
+  & (
+    | {
+      type: 'FunctionDeclaration';
+      declare: false; // explicitly have `declare` false for better type narrowing
+      body: FunctionBody | null; // null in the overload case, BUT maybe this should have a different type
+    }
+    | {
+      type: 'TSDeclareFunction';
+      declare: true;
+      body: null;
+    }
+  );
+
+export type ParamPattern = FormalParameter | TSParameterProperty | FormalParameterRest;
+
+export type FunctionExpression = Span & {
   id: BindingIdentifier | null;
-  expression: false;
   generator: boolean;
   async: boolean;
   params: ParamPattern[];
-  body: FunctionBody | null;
-  declare?: boolean;
   typeParameters?: TSTypeParameterDeclaration | null;
   returnType?: TSTypeAnnotation | null;
-}
-
-export type ParamPattern = FormalParameter | TSParameterProperty | FormalParameterRest;
-
-export type FunctionType =
-  | 'FunctionDeclaration'
-  | 'FunctionExpression'
-  | 'TSDeclareFunction'
-  | 'TSEmptyBodyFunctionExpression';
+  type: 'FunctionExpression';
+  body: FunctionBody; // body always defined for FunctionExpressions
+};
 
 export interface FormalParameterRest extends Span {
   type: 'RestElement';

```
</details>
</collapse>

happy to share my reasoning if needed

**Comment by overlookmotel** [comment]
Would this also work?

1. Leave `Function` type as it is.

2. Add these type definitions:

```ts
export interface FunctionDeclaration extends Function {
    type: 'FunctionDeclaration';
    id: BindingIdentifier;
    body: FunctionBody;
    declare?: false;
}

export interface FunctionExpression extends Function {
    type: 'FunctionExpression';
    body: FunctionBody;
    declare?: false;
}

export interface TSDeclareFunction extends Function {
    type: 'TSDeclareFunction';
    declare?: true;
}

export interface TSEmptyBodyFunctionExpression extends Function {
    type: 'TSEmptyBodyFunctionExpression';
    /* not sure about this one! */
}
```

3. Alter the types of nodes which contain a `Function`:

```ts
export type Expression =
    | FunctionExpression // Not `Function`
    | /* others */;

export type Declaration =
    | FunctionDeclaration // Not `Function`
    | TSDeclareFunction
    | /* others */;
```

Does that achieve the same aim of allowing narrowing based on the `type` field?

If the above would work, it would be easier to achieve without large changes to the codegen.

**Comment by camc314** [comment]
yes that should work, i made some more changes to improve it slighly.

```ts
export interface FunctionDeclaration extends Function {
    type: 'FunctionDeclaration';
    id: BindingIdentifier;
    body: FunctionBody;
    declare?: false;
    expression: false; // can never be true for function declaration
}

export interface FunctionExpression extends Function {
    type: 'FunctionExpression';
    body: FunctionBody;
    declare?: false;
    expression: true; // will always be true
}

export interface TSDeclareFunction extends Function {
    type: 'TSDeclareFunction';
    body: null
    declare?: true;
    expression: false;// will always be false
}

// we can completly remove this one as it's impossible to have a function expression without a body
export interface TSEmptyBodyFunctionExpression extends Function {
    type: 'TSEmptyBodyFunctionExpression';
    /* not sure about this one! */
}
```

3. Alter the types of nodes which contain a `Function`:

```ts
export type Expression =
    | FunctionExpression // Not `Function`
    | /* others */;

export type Declaration =
    | FunctionDeclaration // Not `Function`
    | TSDeclareFunction
    | /* others */;
```

we can remove `TSEmptyBodyFunctionExpression` since it's an impossible shape
we should add `expression: true`/`expression:false` explicitl on the function expression/function declarations

**Comment by Boshen** [comment]
No consensus in 6 months, close as stale.
